### PR TITLE
Run parallel_algorithms with /O2

### PR DESCRIPTION
**Pull request by AlexGuteniev**
To find real-world races

### CODE REVIEW

File: tests/std/tests/parallel_algorithms.lst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier Apache-2.0 WITH LLVM-exception
+
+RUNALL_INCLUDE .\native_17_matrix.lst
```
Comment (CaseyCarter) [inline tests/std/tests/parallel_algorithms.lst:4]: Did you intend to do something here other than adding a layer of indirection? Cross with `/O2`, perhaps?

Comment (AlexGuteniev) [reply tests/std/tests/parallel_algorithms.lst:4]: Oh, yes!

**Review by CaseyCarter — COMMENTED** [review]
I now agree that the change is _correct_, but I'm not completely sold that the change is desirable. Why should we test only this small part of the library with `/O2`? Does doing so really mean we need to double the test load by running every (previous) configuration both with and without `/O2`?

**Comment by AlexGuteniev** [comment]
Most of the othe tests are not multithreaded. Some other multithreaded tests are still of less concern that debug/release timing difference willl show a race

**Comment by StephanTLavavej** [comment]
Some cost data: on my personal machine (5950X, 16C 32T), `main` runs 754 tests from these directories in 83.27s, while this PR runs 1508 tests (as expected, 2x) in 183.40s. That's an extra 100 seconds for a full test pass, which admittedly I don't do all that often.

The impact on CI runs should be roughly 1/8th, since the extra cost is split over 8 shards, and our 32-core VMs are vaguely comparable to my machine (not sure exactly how fast). An extra 12.5 seconds per test run isn't too extreme, although it is certainly a notable increase compared to most PRs. The cost may be more if we ever gain ARM64 execution.

Like Casey, I feel lukewarm about this change. Mostly it's because it's extremely rare for a libraries bug to repro readily under `/O2` but not under `/Od`. (Usually that's the sign of a compiler backend bug.) We don't even typically have bugs of form "bad code that's revealed by optimizations" regardless of timing. And when we did have a runtime correctness bug in the parallel algorithms, our `/Od` runs found it (it was being concealed by our internal test harness's accursed rerun machinery, but the GitHub runs found it repeatedly).

I'm not even sure if `/O2` helps or hurts when it comes to finding data races. Compiler reordering etc. of code can make it more likely, but stuff completing lightning fast could make it less likely (less contention, more likely that stuff finishes before other stuff starts).

I think I'd accept `/O2` with less hesitation if it took our usual form of "take an existing matrix and apply `/O2` to one or a small subset of configurations", instead of crosslisting the whole thing. (This is annoyingly more invasive, but at least it can still be centralized to a single file.)

**Pull request closed without merging**